File: java/demos/ps/oo/City.java
package demos.ps.oo;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;

import demos.ps.oo.PSGame.Planet;
import demos.ps.oo.PSLibMusic.PS1Music;

enum Flyable {YES, NO};

public enum City { 
	CAMINEET	(83, 49, Planet.PALMA, Flyable.YES, PS1Music.TOWN),
	PAROLIT		(73, 58, Planet.PALMA, Flyable.YES, PS1Music.TOWN),
	SCION		(101, 46, Planet.PALMA, Flyable.YES, PS1Music.TOWN),
	EPPI		(85, 80, Planet.PALMA, Flyable.YES, PS1Music.VILLAGE),
	GOTHIC		(53, 58, Planet.PALMA, Flyable.YES, PS1Music.VILLAGE),
	BORTEVO		(25, 51, Planet.PALMA, Flyable.YES, PS1Music.VILLAGE),
	LOAR		(65, 31, Planet.PALMA, Flyable.YES, PS1Music.VILLAGE),
	ABION		(13, 18, Planet.PALMA, Flyable.YES, PS1Music.VILLAGE),
	DRASGOW		(118, 81, Planet.PALMA, Flyable.YES, PS1Music.TOWN),

	SPACEPORT1	(0, 0, Planet.PALMA, Flyable.NO, PS1Music.TOWN),
	SPACEPORT2	(0, 0, Planet.MOTAVIA, Flyable.NO, PS1Music.TOWN),
	
	PASEO		(78, 31, Planet.MOTAVIA, Flyable.YES, PS1Music.TOWN),
	UZO			(93, 64, Planet.MOTAVIA, Flyable.YES, PS1Music.VILLAGE),
	CASBA		(68, 86, Planet.MOTAVIA, Flyable.YES, PS1Music.VILLAGE),
	SOPIA		(18, 32, Planet.MOTAVIA, Flyable.YES, PS1Music.VILLAGE),
	TONOE		(39, 84, Planet.MOTAVIA, Flyable.YES, PS1Music.VILLAGE),

	SKURE_ENTRANCE	(172, 73, Planet.DEZORIS, Flyable.NO, PS1Music.DEZORIS),
	SKURE			(172, 73, Planet.DEZORIS, Flyable.YES, PS1Music.TOWN),
	AUKBA_ENTRANCE	(184, 34, Planet.DEZORIS, Flyable.NO, PS1Music.DEZORIS),
	AUKBA			(184, 34, Planet.DEZORIS, Flyable.YES, PS1Music.VILLAGE), 	
	
	SKY_CASTLE	(0, 0, Planet.PALMA, Flyable.NO, PS1Music.TOWN), 	
	;
	
	private int x, y;
	Planet planet;
	Flyable flyable;
	PS1Music music;
	
	private City(int x, int y, Planet planet, Flyable flyable, PS1Music music) {
		this.x = x;
		this.y = y;
		this.planet = planet;
		this.flyable = flyable;
		this.music = music;
	}
	
	public static List<City> getVisitedCitiesFromPlanet(Planet chosenPlanet, EnumSet<City> visitedCities) {
		List<City> lstCities = new ArrayList<City>();
		for(City city: City.values()) {
			if(city.planet == chosenPlanet && city.flyable == Flyable.YES && 
					(visitedCities == null || visitedCities.contains(city))) {
				lstCities.add(city);
			}
		}
		return lstCities;
	}
	
	@Override
	public String toString() {
		String s = super.toString();
		return PSGame.getString("City_" + s.substring(0, 1) + s.substring(1).toLowerCase());
	}

	public int getX() {
		return x;
	}
	public int getY() {
		return y;
	}
	public PS1Music getMusic() {
		return music;
	}
	public String getPath() {
		String s = this.name().substring(0, 1) + this.name().substring(1).toLowerCase() + ".map";
		return "maps/" + s;
	}
};

File: src/demos/ps/game/City.ts
/**
 * City - City and Location Management
 * Defines all cities and locations in the PS universe
 */

import { PS1Music } from './PSLibMusic';
import { PSGame } from '../PSGame';

export enum Planet {
  PALMA, MOTAVIA, DEZORIS
}

export class PlanetHelper {
  private static readonly planetConfigs = new Map<Planet, { mapPath: string, music: PS1Music }>([
    [Planet.PALMA, { mapPath: "Palma.map", music: PS1Music.PALMA }],
    [Planet.MOTAVIA, { mapPath: "Motavia.map", music: PS1Music.MOTAVIA }],
    [Planet.DEZORIS, { mapPath: "Dezoris.map", music: PS1Music.DEZORIS }]
  ]);

  public static getMapPath(planet: Planet): string {
    return this.planetConfigs.get(planet)?.mapPath || "";
  }

  public static getMusic(planet: Planet): PS1Music {
    return this.planetConfigs.get(planet)?.music || PS1Music.PALMA;
  }
}

enum Flyable {
  YES, NO
}

export enum City {
  // Palma cities and villages
  CAMINEET,
  PAROLIT,
  SCION,
  EPPI,
  GOTHIC,
  BORTEVO,
  LOAR,
  ABION,
  DRASGOW,

  // Spaceports
  SPACEPORT1,
  SPACEPORT2,

  // Motavia cities and villages
  PASEO,
  UZO,
  CASBA,
  SOPIA,
  TONOE,

  // Dezoris locations
  SKURE_ENTRANCE,
  SKURE,
  AUKBA_ENTRANCE,
  AUKBA,

  // Special locations
  SKY_CASTLE
}

export class CityHelper {
  private static readonly cityConfigs = new Map<City, { x: number, y: number, planet: Planet, flyable: Flyable, music: PS1Music }>([
    // Palma cities and villages
    [City.CAMINEET, { x: 83, y: 49, planet: Planet.PALMA, flyable: Flyable.YES, music: PS1Music.TOWN }],
    [City.PAROLIT, { x: 73, y: 58, planet: Planet.PALMA, flyable: Flyable.YES, music: PS1Music.TOWN }],
    [City.SCION, { x: 101, y: 46, planet: Planet.PALMA, flyable: Flyable.YES, music: PS1Music.TOWN }],
    [City.EPPI, { x: 85, y: 80, planet: Planet.PALMA, flyable: Flyable.YES, music: PS1Music.VILLAGE }],
    [City.GOTHIC, { x: 53, y: 58, planet: Planet.PALMA, flyable: Flyable.YES, music: PS1Music.VILLAGE }],
    [City.BORTEVO, { x: 25, y: 51, planet: Planet.PALMA, flyable: Flyable.YES, music: PS1Music.VILLAGE }],
    [City.LOAR, { x: 65, y: 31, planet: Planet.PALMA, flyable: Flyable.YES, music: PS1Music.VILLAGE }],
    [City.ABION, { x: 13, y: 18, planet: Planet.PALMA, flyable: Flyable.YES, music: PS1Music.VILLAGE }],
    [City.DRASGOW, { x: 118, y: 81, planet: Planet.PALMA, flyable: Flyable.YES, music: PS1Music.TOWN }],

    // Spaceports
    [City.SPACEPORT1, { x: 0, y: 0, planet: Planet.PALMA, flyable: Flyable.NO, music: PS1Music.TOWN }],
    [City.SPACEPORT2, { x: 0, y: 0, planet: Planet.MOTAVIA, flyable: Flyable.NO, music: PS1Music.TOWN }],

    // Motavia cities and villages
    [City.PASEO, { x: 78, y: 31, planet: Planet.MOTAVIA, flyable: Flyable.YES, music: PS1Music.TOWN }],
    [City.UZO, { x: 93, y: 64, planet: Planet.MOTAVIA, flyable: Flyable.YES, music: PS1Music.VILLAGE }],
    [City.CASBA, { x: 68, y: 86, planet: Planet.MOTAVIA, flyable: Flyable.YES, music: PS1Music.VILLAGE }],
    [City.SOPIA, { x: 18, y: 32, planet: Planet.MOTAVIA, flyable: Flyable.YES, music: PS1Music.VILLAGE }],
    [City.TONOE, { x: 39, y: 84, planet: Planet.MOTAVIA, flyable: Flyable.YES, music: PS1Music.VILLAGE }],

    // Dezoris locations
    [City.SKURE_ENTRANCE, { x: 172, y: 73, planet: Planet.DEZORIS, flyable: Flyable.NO, music: PS1Music.DEZORIS }],
    [City.SKURE, { x: 172, y: 73, planet: Planet.DEZORIS, flyable: Flyable.YES, music: PS1Music.TOWN }],
    [City.AUKBA_ENTRANCE, { x: 184, y: 34, planet: Planet.DEZORIS, flyable: Flyable.NO, music: PS1Music.DEZORIS }],
    [City.AUKBA, { x: 184, y: 34, planet: Planet.DEZORIS, flyable: Flyable.YES, music: PS1Music.VILLAGE }],

    // Special locations
    [City.SKY_CASTLE, { x: 0, y: 0, planet: Planet.PALMA, flyable: Flyable.NO, music: PS1Music.TOWN }]
  ]);

  /**
   * Get visited cities from planet
   */
  public static getVisitedCitiesFromPlanet(chosenPlanet: Planet, visitedCities?: Set<City>): City[] {
    const lstCities: City[] = [];

    for (const [city, config] of this.cityConfigs) {
      if (config.planet === chosenPlanet &&
          config.flyable === Flyable.YES &&
          (visitedCities === undefined || visitedCities.has(city))) {
        lstCities.push(city);
      }
    }

    return lstCities;
  }

  /**
   * Get all cities
   */
  public static getAllCities(): City[] {
    return Array.from(this.cityConfigs.keys());
  }

  /**
   * Get city X coordinate
   */
  public static getX(city: City): number {
    return this.cityConfigs.get(city)?.x || 0;
  }

  /**
   * Get city Y coordinate
   */
  public static getY(city: City): number {
    return this.cityConfigs.get(city)?.y || 0;
  }

  /**
   * Get city music
   */
  public static getMusic(city: City): PS1Music {
    return this.cityConfigs.get(city)?.music || PS1Music.TOWN;
  }

  /**
   * Get city planet
   */
  public static getPlanet(city: City): Planet {
    return this.cityConfigs.get(city)?.planet || Planet.PALMA;
  }

  /**
   * Check if city is flyable
   */
  public static isFlyable(city: City): boolean {
    return this.cityConfigs.get(city)?.flyable === Flyable.YES;
  }

  /**
   * Get localized city name
   */
  public static toString(city: City): string {
    const s = city.toString();
    const formatted = s.substring(0, 1) + s.substring(1).toLowerCase();
    return PSGame.getString(`City_${formatted}`);
  }

  /**
   * Get map path
   */
  public static getPath(city: City): string {
    const name = city.toString();
    const formatted = name.substring(0, 1) + name.substring(1).toLowerCase() + ".map";
    return `maps/${formatted}`;
  }
}